src/sa.py

```diff
@@ -1,15 +1,194 @@
+import random
+import math
+
+#####################
+# Functions:
+
+from cigar import edits_to_cigar
+from align import get_edits
+    
+# Simulate string:
+def simulate_string(m):
+    """Simulate a DNA sequence of length m.
+    
+    Example:
+    simulate_string(20)
+    
+    """
+    DNA = 'ACGT'
+    nucleotides = [random.choice(DNA) for _ in range(m)]
+    return ''.join(nucleotides)
+    
+
+# Naive approach
+def get_suffix_array(string):
+    string+='$'
+    SA = sorted(range(len(string)), key=lambda i: string[i:])
+    return SA
+
+    
+# Prefix-doubling approach (SIMPLE EDITION - no radix)
+def SuffixArray(string):
+    '''Algorithm for constructing suffix-arrays
+    
+    Example:
+    >>> SuffixArray('trollolol')
+    [9, 8, 3, 6, 4, 7, 2, 5, 1, 0]
+    
+    '''
+    if string == '' or string == None:
+        return None
+    string += '$'
+    index = {v: i for i, v in enumerate(sorted(set(string)))}
+    string = [index[v] for v in string]
+    rank_list = list(range(len(string)))
+    SA = [None]*len(string)
+    tuple_list = SA[:]
+    M,j = 0,1
+    while M < len(string)-1:
+        for i in range(len(string)):
+            i_j = i+j
+            if i_j < len(string):
+                key = (string[i], string[i_j])
+            else: 
+                key = (string[i], string[-1])
+            tuple_list[i] = key
+        j*=2
+        keys = sorted(set(tuple_list))
+        ranks = dict(zip(keys, rank_list))
+        string = [ranks[tuple_list[i]] for i in range(len(string))]
+        M = max(string)
+    for i in rank_list:
+        SA[string[i]] = i
+    return SA
+
+
+def binary_search(SA, string, pattern):
+    '''Algorithm for finding exact patterns from suffix array.
+    
+    Example:
+    string = 'trollolol'
+    pattern = 'ol'
+    SA = SuffixArray(string)
+    binary_search(SA, string, pattern)
+    [7, 2, 5]
+    
+    '''
+    if pattern == '' or pattern == None:
+        return None
+    
+    if string == '' or string == None:
+        return None
+    
+    if SA == '' or SA == None:
+        return None
+    
+    string+='$'
+    SA_positions = []
+    # Binary search.
+    upper, lower = len(SA), 0, 
+    last, match =  None, None
+    j = 0
+    S = 0
+    max_runs = math.ceil(math.log2(len(SA)))
+    while j < max_runs and S == 0:
+        mid = lower + (upper - lower) // 2
+        if mid == last: 
+            S = 1
+        count = 0
+        for i in range(count, len(pattern)):
+            if pattern[i] == string[SA[mid]+i]:
+                count+=1
+            if count == len(pattern):
+                match = mid    
+                SA_positions.append(SA[match])
+                S = 1
+                break
+            elif pattern[i] > string[SA[mid]+i]:
+                    lower = mid
+                    break
+            elif pattern[i] < string[SA[mid]+i]:
+                    upper = mid
+                    break
+        last=mid
+        j+=1 
+    if match != None:

```

Comment:
This looks a lot harder than it has to be. You really only need to do a normal `lower_bound(x, p[i])` search on an interval, where the key is the value at an offset of the suffixes. To find the other end of the interval, you do an upper bound as `lower_bound(x, p[i]+1)`. You then do this iteratively with increasing offsets `i` until `i == len(p)`.

```python
def lower(a: str, offset: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
    while lo < hi:
        m = (lo + hi) // 2
        if x[sa[m] + offset] < a:
            lo = m + 1
        else:
            hi = m
    return lo
```